content/posts/opa.md

```diff
@@ -0,0 +1,555 @@
+This allows having the current OpenStack policies implemented naturally in Rego
+and be extended with the dynamic data and ReBAC style assignments providing a so
+desired freedom.
+
+## OpenPolicyAgent Rego language
+
+As mentioned above Rego is a language used in the OpenPolicyAgent to describe
+the necessary policy behavior. It is not necessary an intuitive language, but
+is also not unnecessary complicated. 
+
+`oslo.policy` based description for listing application credentials:
+
+```yaml
+"identity:list_application_credentials": "(rule:admin_required) or (role:reader and system_scope:all) or rule:owner
+"admin_required": "role:admin or is_admin:1"
+"owner": "user_id:%(user_id)s"
+```
+The following policy permits the API request for admin users (when the user is
+having the "admin" role assigned or the "is_admin" flag set to 1), or the user
+has a "reader" role assigned for the system scoped token or the user is the
+owner of the application credentials (identified with the check comparing the
+"user_id" parameter of the request with the "user_id" of the authenticated
+user). Such policy can be represented in the Rego language as follows:
+
+```
+allow if {
+  # role:admin
+  "admin" in input.credentials.roles
+}
+
+allow if {
+  # role:reader and system_scope:all
+  "reader" in input.credentials.roles
+  input.credentials.system_scope == "all"
+}
+
+allow if {
+  # user_id:%(user_id)s
+  input.credentials.user_id == input.target.user_id
+}
+```
+
+The example above defines 3 rules with the rule header stating "allow" as the
+final authorization decision. The fact that all 3 rules are having the same
+header is having the "OR" effect instructing OPA to permit the request if at
+least one of the rules was evaluated to true.
+
+First rule is ensuring "admin" is present in the credential roles list. The
+second one evaluates to true when credential roles contain "reader" item AND the
+"system_scope" credential attribute is set to "all". Last rule validates that
+the "user_id" attribute of the credential is equal to the "user_id" attribute of
+the target. All this rules are evaluated in parallel and when one of them is
+true, the further evaluation is aborted immediately returning the result.
+
+Let's see how to implement a fine granular access granting the user this rule
+individually without allocation of the new roles. This is where the ReBAC
+shines. It is possible to consider every api operation as a separate target
+that can be accessed by the user though establishing of a direct relation.
+Technically this is not much different to how it is implemented now, since access
+to the target (api operation) is being granted to the user through the roles.
+
+```
+...
+allow if {
+    data.assignments["list_application_credentials"][input.credentials.user_id]
+}
+```
+
+With this addition the request is being permitted by the OPA also if user_id
+entry is present in the supplementary data under the
```

Comment:
Maybe it's worth expanding a bit on this additional data? How and where it can be added so that OPA will consider it?